File: update_discord_topka.py
import json
import os
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timezone


def http_json(url, method="GET", payload=None, timeout=20):
    data = None
    headers = {}
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        raw = resp.read()
        if not raw:
            return {}
        return json.loads(raw.decode("utf-8"))


def build_rows(leaderboard):
    if not isinstance(leaderboard, dict):
        return []
    rows = []
    for nick, cash in leaderboard.items():
        nick = str(nick or "").strip()
        if not nick or nick.lower() == "gracz":
            continue
        try:
            cash_n = float(cash)
        except (TypeError, ValueError):
            continue
        rows.append((nick, cash_n))
    rows.sort(key=lambda x: x[1], reverse=True)
    return rows[:10]


def build_payload(rows):
    medals = ["🥇", "🥈", "🥉"]
    lines = []
    for i, (nick, cash) in enumerate(rows):
        prefix = medals[i] if i < 3 else "•"
        lines.append(f"{prefix} {nick} — {int(cash)}$")
    if not lines:
        lines = ["Brak danych leaderboardu."]
    return {
        "embeds": [
            {
                "title": "💰 Leaderboard",
                "description": "\n".join(lines),
                "color": 0x2ECC71,
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "footer": {"text": "Auto refresh (GitHub Actions)"},
            }
        ]
    }


def main():
    db_url = (os.environ.get("FIREBASE_DB_URL") or "").strip().rstrip("/")
    webhook_url = (os.environ.get("DISCORD_WEBHOOK_URL") or "").strip().rstrip("/")
    message_id = (os.environ.get("DISCORD_MESSAGE_ID") or "").strip()

    if not db_url or not webhook_url or not message_id:
        raise RuntimeError("Brak sekretow: FIREBASE_DB_URL / DISCORD_WEBHOOK_URL / DISCORD_MESSAGE_ID")

    leaderboard_url = f"{db_url}/leaderboard.json"
    leaderboard = http_json(leaderboard_url, method="GET")
    rows = build_rows(leaderboard)
    payload = build_payload(rows)

    edit_url = f"{webhook_url}/messages/{urllib.parse.quote(message_id)}"
    try:
        http_json(edit_url, method="PATCH", payload=payload)
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="ignore")
        raise RuntimeError(f"Discord PATCH error {exc.code}: {body}") from exc

    print(f"OK: zaktualizowano topke, wiersze={len(rows)}")


if __name__ == "__main__":
    main()

File: update_discord_topka.mjs
const env = process.env;

function normalizeWebhook(url) {
  let u = String(url || "").trim().replace(/^['"]|['"]$/g, "");
  u = u.replace("https://discord.com/api/webhooks/", "https://discord.com/api/v10/webhooks/");
  u = u.replace("https://discordapp.com/api/webhooks/", "https://discord.com/api/v10/webhooks/");
  return u.replace(/\/+$/, "");
}

function buildPayload(rows) {
  const medals = ["🥇", "🥈", "🥉"];
  const lines = rows.length
    ? rows.map(([nick, cash], i) => `${medals[i] || "•"} ${nick} — ${Math.floor(cash)}$`)
    : ["Brak danych leaderboardu."];
  return {
    embeds: [
      {
        title: "💰 Leaderboard",
        description: lines.join("\n"),
        color: 0x2ecc71,
        timestamp: new Date().toISOString(),
        footer: { text: "Auto refresh (GitHub Actions)" }
      }
    ]
  };
}

async function readLeaderboard(dbUrl) {
  const resp = await fetch(`${dbUrl.replace(/\/+$/, "")}/leaderboard.json`, {
    headers: { "Accept": "application/json", "User-Agent": "KasynoTopkaBot/2.0" }
  });
  if (!resp.ok) throw new Error(`Firebase GET ${resp.status}: ${await resp.text()}`);
  const data = (await resp.json()) || {};
  const rows = Object.entries(data)
    .map(([nick, cash]) => [String(nick || "").trim(), Number(cash) || 0])
    .filter(([nick]) => nick && nick.toLowerCase() !== "gracz")
    .sort((a, b) => b[1] - a[1])
    .slice(0, 10);
  return rows;
}

async function discordRequest(url, method, payload) {
  const resp = await fetch(url, {
    method,
    headers: {
      "Content-Type": "application/json",
      "Accept": "application/json",
      "User-Agent": "KasynoTopkaBot/2.0"
    },
    body: JSON.stringify(payload)
  });
  if (resp.ok) {
    const txt = await resp.text();
    try { return txt ? JSON.parse(txt) : {}; } catch { return {}; }
  }
  const body = await resp.text();
  throw new Error(`Discord ${method} ${resp.status}: ${body}`);
}

async function main() {
  const dbUrl = String(env.FIREBASE_DB_URL || "").trim();
  const webhookUrl = normalizeWebhook(env.DISCORD_WEBHOOK_URL || "");
  const messageId = String(env.DISCORD_MESSAGE_ID || "").trim();

  if (!dbUrl || !webhookUrl) {
    throw new Error("Brak FIREBASE_DB_URL lub DISCORD_WEBHOOK_URL.");
  }
  if (!webhookUrl.includes("/api/v10/webhooks/")) {
    throw new Error("DISCORD_WEBHOOK_URL nie jest poprawnym webhookiem Discord API.");
  }

  const payload = buildPayload(await readLeaderboard(dbUrl));

  if (messageId) {
    try {
      await discordRequest(`${webhookUrl}/messages/${encodeURIComponent(messageId)}`, "PATCH", payload);
      console.log("OK: PATCH leaderboard success");
      return;
    } catch (e) {
      console.log(`PATCH failed -> fallback POST: ${e.message}`);
    }
  }

  const created = await discordRequest(`${webhookUrl}?wait=true`, "POST", payload);
  const newId = String((created && created.id) || "").trim();
  if (!newId) throw new Error("POST ok, ale brak id wiadomosci w odpowiedzi.");
  console.log(`OK: stworzono nowa wiadomosc. Ustaw DISCORD_MESSAGE_ID=${newId}`);
}

main().catch((err) => {
  console.error(err.message || String(err));
  process.exit(1);
});
